internal/server/v2/observation/calculation_util.go

```diff
@@ -191,3 +195,105 @@ func findObservationResponseHoles(
 	}
 	return result, nil
 }
+
+// Find all candidate series that match each ASTNode and add to VariableFormula.
+func computeLeafSeries(
+	inputResp *pbv2.ObservationResponse,
+	formula *VariableFormula,
+) {
+	for _, leafData := range formula.LeafData {
+		leafData.CandidateSeries = map[string]map[string][]*pb.PointStat{}
+		variableObs, ok := inputResp.ByVariable[leafData.StatVar]
+		// No data for input variable.
+		if !ok {
+			return
+		}
+		for entity, entityObs := range variableObs.ByEntity {
+			facetMap := map[string][]*pb.PointStat{}
+			for _, facetObs := range entityObs.OrderedFacets {
+				if leafData.Facet != nil {
+					facet := inputResp.Facets[facetObs.FacetId]
+					if leafData.Facet.GetMeasurementMethod() != facet.GetMeasurementMethod() {
+						continue
+					}
+					if leafData.Facet.GetObservationPeriod() != facet.GetObservationPeriod() {
+						continue
+					}
+					if leafData.Facet.GetUnit() != facet.GetUnit() {
+						continue
+					}
+					if leafData.Facet.GetScalingFactor() != facet.GetScalingFactor() {
+						continue
+					}
+				}
+				facetMap[facetObs.FacetId] = facetObs.Observations
+			}
+			if len(facetMap) > 0 {
+				leafData.CandidateSeries[entity] = facetMap
+			}
+		}
+	}
+}
+
+// Combine two sets of candidate series in one step of calculation.
```

Comment:
"one step of calculation" is a bit confusing.. It feels like calculation needs multiple steps

Reply:
Updated the comment to be more explicit